modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileFeatureSource.java

```diff
@@ -382,12 +370,8 @@ protected int getCountInternal(Query query) throws IOException {
             reader.setFilter(filter);
         }
 
-        // setup the target bbox if any, and the generalization hints if available
+        // setup the generalization hints if available
         if (q != null) {
-            if (bbox != null && !bbox.isNull()) {
```

Comment:
This change will cause a severe slowdown for non spatially indexed shapefiles, as the reader can skip a feature if the bbox of its geometry does not intersect the requested bbox. But if the bbox is not set anymore, this won't be possible anymore.
The regression will also happen for indexed files, as the quadtree is compact and leads to false positives.

Long story short, this change one needs to be reverted, compute the bbox anyways and use it only to set it in the reader.